src/types.rs

```diff
@@ -157,10 +166,130 @@ impl std::fmt::Display for InstructionForm {
     }
 }
 
+// Register ---------------------------------------------------------------------------------------
+#[derive(Debug, Clone)]
+pub struct Register {
+    pub name: String,
+    pub alt_names: Vec<String>,
+    pub description: Option<String>,
+    pub reg_type: Option<RegisterType>,
+    pub width: Option<RegisterWidth>,
+    pub flag_info: Vec<RegisterBitInfo>,
+    pub arch: Option<Arch>,
+    pub url: Option<String>,
+}
+
+impl Hoverable for &Register {}
+
+impl Default for Register {
+    fn default() -> Self {
+        let name = String::new();
+        let alt_names = vec![];
+        let description = None;
+        let reg_type = None;
+        let width = None;
+        let flag_info = vec![];
+        let arch = None;
+        let url = None;
+
+        Self {
+            name,
+            alt_names,
+            description,
+            reg_type,
+            width,
+            flag_info,
+            arch,
+            url,
+        }
+    }
+}
+
+impl std::fmt::Display for Register {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        // basic fields
+        let header: String;
+        if let Some(arch) = &self.arch {
+            header = format!("{} [{}]", &self.name.to_uppercase(), arch.as_ref());
+        } else {
+            header = self.name.to_uppercase();
+        }
+
+        let mut v: Vec<&str> = if let Some(description_) = &self.description {
+            vec![&header, description_, "\n"]
+        } else {
+            vec![&header, "\n"]
+        };
+
+        // Register Type
+        let reg_type;
+        if let Some(reg_type_) = &self.reg_type {
+            reg_type = format!("Type: {}", reg_type_);
+            v.push(reg_type.as_str());
+        }
+
+        // Register Width
+        let reg_width;
```

Comment:
Can  this be defined inside the `if let`?

Reply:
I originally tried that, but if it's defined inside the ``if let`` the variable then gets dropped immediately after the ``if let``'s scope ends.
![image](https://github.com/bergercookie/asm-lsp/assets/87077023/160a8536-1be6-4016-8a06-0b7cdaeeeaf6)

Reply:
Yes, that's because the vector `v` holds `&str`s instead of `String`. I think it would be cleaner to instead hold `String`, and define them inside the conditional. This way the rust compiler will std::move them to the vector instead of just copying a reference to them.

Here's the changes I had in mind: https://github.com/WillLillis/asm-lsp/pull/2
That's more of a preference though, not a big deal either way.

Reply:
Oh, I see. Looks good to me!